GVFS/GVFS.Common/Tracing/QueuedPipeStringWriter.cs

```diff
@@ -0,0 +1,186 @@
+
+namespace GVFS.Common.Tracing
+{
+    public enum QueuedPipeStringWriterState
+    {
+        Unknown = 0,
+        Stopped = 1,
+        Failing = 2,
+        Healthy = 3,
+    }
+
+    /// <summary>
+    /// Accepts string messages from multiple threads and dispatches them over a named pipe from a
+    /// background thread.
+    /// </summary>
+    public class QueuedPipeStringWriter : IDisposable
+    {
+        private const int DEFAULT_MAX_QUEUE_SIZE = 256;
+
+        private readonly Func<NamedPipeClientStream> createPipeFunc;
+        private readonly IQueuedPipeStringWriterEventSink eventSink;
+        private readonly BlockingCollection<string> queue;
+
+        private Thread writerThread;
+        private NamedPipeClientStream pipeClient;
+        private QueuedPipeStringWriterState state = QueuedPipeStringWriterState.Unknown;
+        private bool isDisposed;
+
+        public QueuedPipeStringWriter(Func<NamedPipeClientStream> createPipeFunc, IQueuedPipeStringWriterEventSink eventSink, int maxQueueSize = DEFAULT_MAX_QUEUE_SIZE)
+        {
+            this.createPipeFunc = createPipeFunc;
+            this.eventSink = eventSink;
+            this.queue = new BlockingCollection<string>(new ConcurrentQueue<string>(), boundedCapacity: maxQueueSize);
+        }
+
+        public void Start()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.writerThread != null)
+            {
+                return;
+            }
+
+            this.writerThread = new Thread(this.ThreadProc)
+            {
+                Name = nameof(QueuedPipeStringWriter),
+                IsBackground = true,
+            };
+
+            this.writerThread.Start();
+        }
+
+        public bool TryEnqueue(string message)
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            return this.queue.TryAdd(message);
+        }
+
+        public void Stop()
+        {
+            if (this.isDisposed)
+            {
+                throw new ObjectDisposedException(nameof(QueuedPipeStringWriter));
+            }
+
+            if (this.queue.IsAddingCompleted)
+            {
+                return;
+            }
+
+            // Signal to the queue draining thread that it should drain once more and then terminate.
+            this.queue.CompleteAdding();
+            this.writerThread.Join();
+
+            Debug.Assert(this.queue.IsCompleted, "Message queue should be empty after being stopped");
+        }
+
+        public void Dispose()
+        {
+            if (!this.isDisposed)
+            {
+                this.Stop();
+
+                this.pipeClient?.Dispose();
+                this.pipeClient = null;
+
+                this.writerThread = null;
+
+                this.queue.Dispose();
```

Comment:
I noticed we're setting `this.pipeClient` to `null`, but not `this.queue`.  Should `this.queue` be set to `null` as well?

Reply:
`this.queue` is currently `readonly` so it cannot be set to `null` without removing this.

In practice the GC will still collect the queue object just as quickly even though this reference still exists since once the reference to the `QueuedPipeStringWriter` is gone. Also given this will be used by the `TelemetryDaemonEventListener` which lives for the lifetime of a `JsonTracer`, which itself lives for the lifetime of the process means these are all going to be Generation 2 objects so are less likely to be reclaimed until a full collection happens.. plus if we're being disposed it means the process is being shutdown in this case.

Reply:
> this.queue is currently readonly so it cannot be set to null without removing this.

Thanks I missed that detail!